databroker/experimental/server_ext.py

```diff
@@ -0,0 +1,501 @@
+        self.sorting = sorting or [("metadata.scan_id", 1)]
+        self.metadata = metadata or {}
+        self.principal = principal
+        self.access_policy = access_policy
+        super().__init__()
+
+    @classmethod
+    def from_uri(cls, uri, directory, *, metadata=None):
+        if not pymongo.uri_parser.parse_uri(uri)["database"]:
+            raise ValueError(
+                f"Invalid URI: {uri!r} " f"Did you forget to include a database?"
+            )
+        client = pymongo.MongoClient(uri)
+        database = client.get_database()
+        return cls(database=database, directory=directory, metadata=metadata)
+
+    @classmethod
+    def from_mongomock(cls, directory, *, metadata=None):
+        import mongomock
+
+        db_name = f"temp-{str(uuid.uuid4())}"
+        mongo_client = mongomock.MongoClient()
+        database = mongo_client[db_name]
+
+        return cls(database=database, directory=directory, metadata=metadata)
+
+    def new_variation(
+        self,
+        metadata=UNCHANGED,
+        queries=UNCHANGED,
+        sorting=UNCHANGED,
+        principal=UNCHANGED,
+        **kwargs,
+    ):
+        if metadata is UNCHANGED:
+            metadata = self.metadata
+        if queries is UNCHANGED:
+            queries = self.queries
+        if sorting is UNCHANGED:
+            sorting = self.sorting
+        if principal is UNCHANGED:
+            principal = self.principal
+        return type(self)(
+            database=self.database,
+            directory=self.directory,
+            metadata=metadata,
+            queries=queries,
+            sorting=sorting,
+            access_policy=self.access_policy,
+            principal=principal,
+            **kwargs,
+        )
+
+    def post_metadata(self, metadata, structure_family, structure, specs):
+
+        mime_structure_association = {
+            StructureFamily.array: "application/octet-stream",
+            StructureFamily.dataframe: APACHE_ARROW_FILE_MIME_TYPE,
+        }
+
+        uid = str(uuid.uuid4())
+
+        validated_document = Document(
+            uid=uid,
+            structure_family=structure_family,
+            structure=structure,
+            metadata=metadata,
+            specs=specs,
+            mimetype=mime_structure_association[structure_family],
+        )
+
+        # After validating the document must be encoded to bytes again to make it compatible with MongoDB
+        if validated_document.structure_family == StructureFamily.dataframe:
+            validated_document.structure.micro.meta = bytes(
+                serialize_arrow(validated_document.structure.micro.meta, {})
+            )
+
+        self.collection.insert_one(validated_document.dict())
+        return uid
+
+    def authenticated_as(self, identity):
+        if self.principal is not None:
+            raise RuntimeError(f"Already authenticated as {self.principal}")
+        if self.access_policy is not None:
```

Comment:
this probably doesn't need to be a blocker but it is probably worth thinking about how access policies should be integrated with the writing methods

Reply:
Yep, let's handle this in a follow-up PR.